<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>CentriMo - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/centrimo_icon.png" alt="CentriMo Logo">
      <h1>CentriMo</h1>
      <h2>Local Motif Enrichment Analysis</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>centrimo [options] &lt;primary sequence file&gt; &lt;motif file&gt;+</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
       <span class="man blurb" data-id="centrimo"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <h3>Input</h3>
      <div class="indent">
        <h4>Motif File(s)</h4>
        <div class="indent">
          <p>File(s) containing <a href="meme-format.html">MEME formatted motifs</a>.
          Outputs from MEME and DREME are supported, as well as Minimal MEME
          Format. You can convert many other motif formats to MEME format
          using <a href="overview.html#motif_conversion_utilities">conversion scripts</a>
          available with the MEME Suite.</p>
        </div>

        <h4>Sequence File</h4>
        <div class="indent">
          <p>A file containing <a href="fasta-format.html">FASTA formatted sequences</a>,
          ideally all of the same length. The sequences in this file are referred to as the
          "primary sequences" when a second set of (control) sequences is provided using
          the <span class="popt">--neg</span> option (see below).</p>
        </div>
      </div>

      <h3>Output</h3>
      <div class="indent">
        <p>CentriMo outputs an HTML file named 
        <a href=examples/centrimo_example_output_files/centrimo.html><code>centrimo.html</code></a>
	that allows interactive selection of which motifs to plot the 
	positional distribution for and control over
        smoothing and other plotting parameters.  CentriMo also
        outputs two text files: 
        <a href=examples/centrimo_example_output_files/centrimo.txt><code>centrimo.txt</code></a>, 
	a tab delimited version of the results, and 
        <a href=examples/centrimo_example_output_files/site_counts.txt><code>site_counts.txt</code></a>, 
	which lists, for each motif and each offset, the number of sequences where the best match of the motif
        occurs at the given offset.</p>
        
      </div>
      <h3>Options</h3>
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th>
              <th>Parameter</th>
              <th>Description</th>
              <th>Default Behaviour</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="4">Input/Output</th>
            </tr>
            <script>
              wropt("all-o", {"ddash": true, "dir": "centrimo_out"});
              wropt("all-oc", {"ddash": true, "dir": "centrimo_out"});
            </script>
            <tr>
              <td class="popt">--neg</td><td><span class="pdat">control sequence file</span></td>
              <td>Plot the motif distributions in this set (the control sequences) as well.</td>
              <td>&nbsp</td>
            </tr>
            <tr>
              <td class="popt">--disc</td>
              <td>&nbsp;</td>
              <td>For each enriched region in the primary sequences, the signficance of the 
              <b>relative</b> enrichment of the motif in that region in the primary
              versus control sequences is evaluated using Fisher's exact test.
              Requires the control sequences to be supplied with the <span class="popt">--neg</span> option.
              <td>Use the binomial test on the <b>primary sequences</b> to evaluate motif enrichment.</td>
            </tr>
            <script>
              wropt("all-xalph", {"ddash": true});
            </script>
            <script>
              wropt("all-bg", {"ddash": true, "bfile":true, "centrimo": true, "sequences": true});
            </script>
            <tr>
              <td class="popt">--motif-pseudo</td>
              <td><span class="pdat">pseudocount</span></td>
              <td>Add a this total pseudocount to the counts in each motif column
                when converting a frequency matrix to a log-odds score matrix. 
	        The pseudocount added to each count is <span class="pdat">pseudocount</span>
		times the background frequency of the letter (see option 
		<span class="popt">--bfile</span>, above).
                <b>Note:</b> Counts are computed from
                <a href=meme-format.html>MEME formatted motifs</a> by multiplying the
                the frequency of the letter times the value of <code>nsites</code> given
		in the motif <code>letter-probability matrix</code> header line.
              </td>
              <td>The program applies a pseudocount of 0.1.</td>
            </tr>
            <tr>
              <td class="popt">--motif</td>
              <td><span class="pdat">ID</span></td>
              <td>Select the motif with the <span class="pdat">ID</span> for scanning.
                This option may be repeated to select multiple motifs.</td>
              <td>The program scans with all the motifs.</td>
            </tr>
            <tr>
              <td class="popt">--seqlen</td><td><span class="pdat">length</span></td>
              <td>Use sequences with the length <span class="pdat">length</span> 
              ignoring all other sequences in the input file(s).</td>
              <td>Use sequences with the same length as the first sequence,
              ignoring all other sequences in the input file(s).</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Scanning</th>
            </tr>
            <tr>
              <td class="popt">--score</td> 
              <td><span class="pdat">S</span></td>   
              <td>The score threshold for PWMs, in bits. Sequences without a match 
              with score &ge; <span class="pdat">S</span> are ignored.</td> 
              <td>A score of 5 is used.</td>
            </tr>
            <tr>
              <td class="popt">--optimize_score</td><td>&nbsp;</td>
              <td>Search for the optimal score above the minimum threshold given by
              the <span class="popt">--score</span> option.</td>
              <td>The minimum score threshold is used.</td>
            </tr>
            <tr>
              <td class="popt">--maxreg</td> 
              <td><span class="pdat">max region</span></td>   
              <td>The maximum region size to consider.</td> 
              <td>Try all region sizes up to the sequence width.</td>
            </tr>
            <tr>
              <td class="popt">--minreg</td> 
              <td><span class="pdat">min region</span></td>   
              <td>The minimum region size to consider.  Must be less than <span class="pdat">max region</span>.</td> 
              <td>Try regions 1 bp and larger.</td>
            </tr>
            <tr>
              <td class="popt">--norc</td> 
              <td>&nbsp;</td>   
              <td>Do not scan with the reverse complement motif.</td> 
              <td>Scans with the reverse complement motif.</td>
            </tr>
            <tr>
              <td class="popt">--flip</td> 
              <td>&nbsp;</td>   
              <td>reverse complement matches appear 'reflected' around sequence centers.</td>
              <td>Do not 'flip' the sequence; use rc of motif instead.</td>
            </tr>
            <tr>
              <td class="popt">--local</td>
              <td>&nbsp;</td>
              <td>Compute enrichment of all regions.</td>
              <td>Compute enrichment of central regions.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Output filtering</th>
            </tr>
            <tr>
              <td class="popt">--ethresh</td>
              <td><span class="pdat">thresh</span></td>
              <td>Limit the results to motifs with an enriched region whose
              <i>E</i>-value is less than <span class="pdat">thresh</span>.
              Enrichment <i>E</i>-values are computed by first adjusting the binomial
              <i>p</i>-value of a region for the number of regions tested using the
              Bonferroni correction, and then multiplying the adjusted <i>p</i>-value
              by the number of motifs in the input to CentriMo.</td>  
              <td>Include motifs with <i>E</i>-values up to 10.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Miscellaneous</th>
            </tr>
            <tr>
              <td class="popt">--desc</td><td><span class="pdat">description</span></td>
              <td>Include the text <span class="pdat">description</span> in the HTML
              output.</td>
              <td>No description in the HTML output.</td>
            </tr>
            <tr>
              <td class="popt">--dfile</td><td><span class="pdat">desc file</span></td>
              <td>Include the first 500 characters of text from the file
              <span class="pdat">desc file</span> in the HTML output.</td>
              <td>No description in the HTML output.</td>
            </tr>
            <tr>
              <td class="popt">--noseq</td>
              <td>&nbsp;</td>
              <td>Do not store sequence IDs in the output of CentriMo.</td>
              <td>CentriMo stores a list of the sequence IDs with matches in the
              best region for each motif. This can potentially make the file
              size much larger.</td>
            </tr>
            <script>wropt("all-verbosity");</script>
            <tr>
              <td class="popt">--version</td><td></td>
              <td>Display the version and exit.</td>
              <td>Run as normal.</td>
            </tr>
          </tbody>
        </table>
      </div>

    </div>

    <h3>Citing</h3>
    <div class="indent">
      <p>If you use CentriMo in your research, please cite the following paper:<br />
        <span class="citation">
          Timothy L. Bailey and Philip Machanick, 
          &quot;Inferring direct DNA binding from ChIP-seq&quot;, 
          <i>Nucleic Acids Research</i>, <b>40</b>:e128, 2012.
          <a href="http://nar.oxfordjournals.org/content/40/17/e128">[full text]</a>
        </span>
      </p>
    </div>

    <script src="js/manual.js"></script>

  </body>
</html>

