<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>DREME - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/dreme_icon.png" alt="DREME Logo">
      <h1>DREME</h1>
      <h2>Discriminative Regular Expression Motif Elicitation</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage (Python 2.7):</h3>
        <code>dreme [options] -p &lt;primary sequence file&gt; [-n &lt;control sequence file&gt;]</code>
        <p />
        <h3>Usage (Python 3.x):</h3>
        <code>dreme-py3 [options] -p &lt;primary sequence file&gt; [-n &lt;control sequence file&gt;]</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
       <span class="man blurb" data-id="dreme"></span>
       <p />
       Two different versions of Dreme are provided: <code>dreme</code> for Python 2.7 and
       <code>dreme-py3</code> for Python 3.x. The random number generator has changed between Python
       2.7 and Python 3.x, so results between the two verisons may differ if
       you don't explicitly provide a control sequence file.
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <h3>Input</h3>
      <div class="indent">
        <h4>Sequence file (primary)</h4>
        <p>A collection of sequences in <a href="fasta-format.html">FASTA</a>
        format. The sequences should all be approximately the same length.</p>
      </div>
      <h3>Output</h3>
      <div class="indent">
        <p>DREME outputs its results primarily as an HTML file named
        <a href=examples/dreme_example_output_files/dreme.html><code>dreme.html</code></a>.
        DREME also outputs a machine-readable XML file and a plain-text versions of its results,
        named <a href=examples/dreme_example_output_files/dreme.xml><code>dreme.xml</code></a>
        and <a href=examples/dreme_example_output_files/dreme.txt><code>dreme.txt</code></a>,
        respectively.</p>
        <p>Additionally DREME can output motif logos if the -png and/or -eps options
        are specified.</p>
      </div>

      <h3>Options</h3> 
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th><th>Parameter</th><th>Description</th><th>Default Behaviour</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="4">Input/Output</th>
            </tr>
            <script>
              wropt("all-o", {"ddash": false, "dir": "dreme_out"});
              wropt("all-oc", {"ddash": false, "dir": "dreme_out"});
            </script>
            <tr>
              <td class="popt">-p</td><td><span class=pdat>primary sequence file</span></td>
              <td>The name of a file containing the primary sequences in 
                <a href="fasta-format.html">FASTA format</a>.
              <td>Required argument.<td>
            </tr>
            <tr>
              <td class="popt">-n</td><td><span class=pdat>control sequence file</span></td>
              <td>The name of a file containing the control sequences in 
                <a href="fasta-format.html">FASTA format</a>.
              <td>DREME will create a set of control sequences by shuffling the primary sequences
                while preserving dimer frequencies.<td>
            </tr>
            <tr>
              <td class="popt">-png</td><td>&nbsp;</td>
              <td>Output motif logo images in portable network graphics (png)
              format. This format is useful for display on websites.</td>
              <td>Images are not output in png format.</td>
            </tr>
            <tr>
              <td class="popt">-eps</td><td>&nbsp;</td>
              <td>Output motif logo images in Encapsulated Postscript (eps)
              format. This format is useful for inclusion in publications as
              it is a vector graphics format and can be easily scaled.</td>
              <td>Images are not output in eps format.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Alphabet</th>
            </tr>
            <tr>
              <td class="popt">-dna</td><td>&nbsp;</td>
              <td>Use the standard DNA alphabet.
                This is the default alphabet anyway so the option only exists for symmetry.</td>
              <td>The standard DNA alphabet is used.</td>
            </tr>
            <tr>
              <td class="popt">-rna</td><td>&nbsp;</td>
              <td>Use the standard RNA alphabet.</td>
              <td>The standard DNA alphabet is used.</td>
            </tr>
            <tr>
              <td class="popt">-protein</td><td>&nbsp;</td>
              <td>Use the standard protein alphabet. This does not work very 
                well because the protein alphabet only has 3 ambiguous symbols
                which does not cover the range of possibilities very well at all.</td>
              <td>The standard DNA alphabet is used.</td>
            </tr>
            <script>
              wropt("all-alph", {"ddash": false, "dreme": true});
            </script>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">General</th>
            </tr>
            <tr>
              <td class="popt">-norc</td><td>&nbsp;</td>
              <td>Search only the given primary sequences for motifs.</td>
              <td>Search the given primary sequences and their reverse
                complements for motifs when the alphabet is complementable.</td>
            </tr>
            <tr>
              <td class="popt">-g</td><td><span class="pdat">ngen</span></td>
              <td>Set the number of REs to generalize. Increasing
              <span class="pdat">ngen</span> will make the search more
              through at some cost in speed.</td>
              <td>DREME will generalise 100 REs.</td>
            </tr>
            <tr>
              <td class="popt">-s</td><td><span class="pdat">seed</span></td>
              <td>Set the seed for the random number generator used to shuffle
              the sequences. When the <span class="popt">-n</span> option is
              given the control sequences will be used instead of shuffling.</td>
              <td>The random number generator is initilised with a value of 1.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Stopping Conditions</th>
            </tr>
            <tr>
              <td class="popt">-e</td><td><span class="pdat">e</span></td>
              <td>Stop searching for motifs when the next motif's <i>E</i>-value
                is &gt; <span class="pdat">e</span>.</td>
              <td>Stop discovering motifs if the <i>E</i>-value threshold of 0.05
              is exceeded.</td>
            </tr>
            <tr>
              <td class="popt">-m</td><td><span class="pdat">m</span></td>
              <td>Stop searching for motifs when <span class="pdat">m</span>
              motifs have been found.</td>
              <td>There is no limit on the number of discovered motifs.</td>
            </tr>
            <tr>
              <td class="popt">-t</td><td><span class="pdat">t</span></td>
              <td>Stop searching for motifs when <span class="pdat">t</span>
              seconds have elapsed.</td>
              <td>There is no limit on the time taken to discover motifs.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Set Core Motif Width</th>
            </tr>
            <tr>
              <td class="popt">-mink</td><td><span class="pdat">mink</span></td>
              <td>Set the minimum width of the motif core.</td>
              <td>A minimum core width of 3 is used.</td>
            </tr>
            <tr>
              <td class="popt">-maxk</td><td><span class="pdat">maxk</span></td>
              <td>Set the maximum width of the motif core.</td>
              <td>A maximum core width of 7 is used.</td>
            </tr>
            <tr>
              <td class="popt">-k</td><td><span class="pdat">k</span></td>
              <td>Set the width of the motif core. This sets minimum width and
              maximum width to the same number.</td>
              <td>The defaults for minimum and maximum width are used.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Miscellaneous</th>
            </tr>
            <tr>
              <td class="popt">-desc</td>
              <td><span class="pdat">description</span></td>
              <td>Include the text <span class="pdat">description</span> in the
              HTML output.</td>
              <td>No description in the HTML output.</td>
            </tr>
            <tr>
              <td class="popt">-dfile</td>
              <td><span class="pdat">desc file</span></td>
              <td>Include the first 500 characters of text from the file
              <span class="pdat">desc file</span> in the HTML output.</td>
              <td>No description in the HTML output.</td>
            </tr>
            <script>wropt("all-verbosity");</script>
            <tr>
              <td class="popt">-h</td><td>&nbsp;</td>
              <td>Display a usage message and exit.</td>
              <td>Run as normal</td>
            </tr>
            <tr>
              <td class="popt">-version</td><td></td>
              <td>Display the version and exit.</td>
              <td>Run as normal.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Experimental (use at own risk)</th>
            </tr>
            <tr>
              <td class="popt">-l</td><td>&nbsp;</td>
              <td>Print list of enrichment of all REs tested.</td>
              <td>&nbsp;</td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>

    <h3>Citing</h3>
    <div class="indent">
      <p>
        If you use DREME in your research please cite the following paper:<br />
        <span class="citation">
          Timothy L. Bailey, "DREME: Motif discovery in transcription factor ChIP-seq data", <i>Bioinformatics</i>, <b>27</b>(12):1653-1659, 2011.
          <a href="http://bioinformatics.oxfordjournals.org/content/early/2011/05/04/bioinformatics.btr261.full">[full text]</a>
        </span>
      </p>
    </div>

    <script src="js/manual.js"></script>

  </body>
</html>

