<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>getsize - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>

    <h2>getsize</h2>
    
    <div class="usage">
      <h3>Usage:</h3>
      <code>getsize &lt;sequences&gt; [options]</code>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <p>The program getsize prints statistics about sequences read from a
      <a href="fasta-format.html">FASTA file</a>.</p>
      <p>When counting letters, alias symbols are first converted to their core symbol.</p>
    </div>

    <h3>Input</h3>
    <div class="indent">
      <h4>Sequences</h4>
      <p>A file in <a href="fasta-format.html">FASTA format</a>.
      If the filename is given as "-", then the program reads from
      standard input.</p>
    </div>

    <h3>Output</h3>
    <div class="indent">
      <p>Writes statistics about the sequences to standard output.</p>
    </div>

    <h3>Options</h3>
    <div class="indent">
      <table class="options">
        <thead>
          <tr>
            <th>Option</th>
            <th>Parameter</th>
            <th>Description</th>
            <th>Default Behaviour</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th colspan="4">General Options</th>
          </tr>
          <tr>
            <td class="popt">-dna</td><td>&nbsp;</td>
            <td>Assume sequences are DNA and print letter frequencies in <a href="bfile-format.html">background file format</a>.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-rna</td><td>&nbsp;</td>
            <td>Assume sequences are RNA and print letter frequencies in <a href="bfile-format.html">background file format</a>.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-prot</td><td>&nbsp;</td>
            <td>Assume sequences are protein and print letter frequencies in <a href="bfile-format.html">background file format</a>.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-alph</td><td><span class="pdat">file</span></td>
            <td>Use the <a href="alphabet-format.html">alphabet defintion</a>
              in the specified <span class="pdat">file</span> and print
              frequencies in <a href="bfile-format.html">background file format</a>.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-f</td><td>&nbsp;</td>
            <td>Print the letter frequences as a C array.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-ft</td><td>&nbsp;</td>
            <td>Print letter frequences in a LaTex table.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-l</td><td>&nbsp;</td>
            <td>Just print the length of each sequence.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-nd</td><td>&nbsp;</td>
            <td>Do not print warnings about duplicate sequences.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-x</td><td>&nbsp;</td>
            <td>Translate DNA in 6 frames (use with <span class=popt>-f</span> or <span class=popt>-ft</span>)
               and print the protein letter frequencies in one C array for each frame.</td>
            <td></td>
          </tr>
          <tr>
            <td class="popt">-codons</td><td>&nbsp;</td>
            <td>As for <span class=popt>-x</span>, and also print frame 0 codon usage
		as a C array.</td>
            <td></td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>

