<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MCAST - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/mcast_icon.png" alt="MCAST Logo">
      <h1>MCAST</h1>
      <h2>Motif Cluster Alignment and Search Tool</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
	<code>mcast [options] &lt;motifs&gt; &lt;sequence database&gt;</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <span class="man blurb" data-id="mcast"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>

      <div class="indent">
	<p>In order for MCAST to compute statistical confidence estimates, at
	least 200 matches must be found. If the database contains too few
	sequences, or if certain other options are made too stringent, then too
	few matches may exist for significance statistics to be computed. In
	this case, the <i>p</i>-value, <i>q</i>-value, and <i>E</i>-value
	columns are set to "NaN", and all matches are printed. This limitation
	can be overcome by specifying the <span class="popt">--synth</span> option.
  When this option is set,
	synthetic sequences will be generated using a background model
  generated by choosing a random GC frequency within the range of
  observed GC minimum and maximum.
	The synthetic sequences will be used to estimate significance statistics.</p>

	<p>When computing statistical confidence estimates, MCAST must retain
	the matches in memory until the final distribution of scores can be
	estimated. This means that the scanning of genome sized datasets has the
	potential to exhaust all available memory. To avoid this problem, MCAST
	uses reservoir sampling of the match scores, and limits the number of
	matches that are kept in memory. The default number of matches kept in
	memory is 100,000, but this value can be adjusted via the
	<span class="popt">--max-stored-scores</span> option. If the maximum
	number of stored matches is reached, then MCAST will drop the least
	significant half of the matches. This behavior may result in matches
	missing from the MCAST output, even though they would have satisfied
	the user-specified <i>p</i>-value or <i>q</i>-value threshold.</p>

	<p>MCAST can make use of position-specific priors (PSPs) to improve its
	identification of true motif occurrences. To take advantage of PSPs in
	MCAST you use must provide two command line options. The
	<span class="popt">--psp</span> option is used to set the name of a
	file containing the PSP,
	and the
	<span class="popt">--prior-dist</span> option is used to set the name of
	a file containing the binned distribution of the PSP.</p>

	<p>
	The PSP can be provided in
	<a href="psp-format.html">MEME PSP file format</a>,
	or in 
	<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">wiggle format</a>.
	The <a href="psp-format.html">MEME PSP file format</a> requires that a
	PSP be included for every position in the sequence to be scanned.
	This format is usually only practical for relatively small sequence
	databases.
	The
	<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">wiggle format</a>
	accommodates sequence segments with missing PSP.
	When no PSP is available for a given position, MCAST will use the median PSP from the PSP
	distribution file.
	The
	<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">wiggle format</a>
	will work with large sequence databases, including full genomes.</p>

	<p>
	The PSP and PSP distribution files can be generated from raw scores using the 
	<a href="create-priors.html"><code>create-priors</code></a> utility available
	when you download and install the MEME Suite on your own computer.

	<p style="margin-bottom:0">A full description of the algorithm may be found in:</p>
	<div class="citation">
	  <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">
	  Bailey and Noble. "Searching for statistically significant regulatory
	  modules." <i>Bioinformatics (Proceedings of the European Conference on
	  Computational Biology)</i>. 19(Suppl. 2):ii16-ii25, 2003.</a>
	</div>
      </div>

      <h3>Input</h3>
      <div class="indent">
	<h4>Motifs</h4>
	<div class="indent">
	  <p>A file containing DNA motifs in <a href="meme-format.html">MEME formatted</a>.
	  Outputs from MEME and DREME are supported, as well as Minimal MEME
	  Format. You can also input DNA motifs in TRANSFAC format if you 
	  specify the <code>--transfac</code>.  You can convert many other motif formats to MEME format
	  using <a href="overview.html#motif_conversion_utilities">conversion scripts</a>
	  available with the MEME Suite.  Input motifs that are likely to appear in the
	  sequences.</p>
	</div>

	<h4>Sequence Database</h4>
	<div class="indent">
	  <p>A collection of DNA sequences in
	  <a href="fasta-format.html">FASTA</a> format.</p>
	</div>
      </div>

      <h3>Output</h3>
      <div class="indent">
	<p style="margin-bottom:0">MCAST will create a directory named
	<code>mcast_out</code> (the name of this directory can be overridden via the
	<span class="popt">--o</span> or <span class="popt">--oc</span> options)
	The directory will contain:</p>
	<ul>
	  <li>A file named <a href=examples/mcast_example_output_files/mcast.html><code>mcast.html</code></a>
	  reporting the matches in HTML format (see details <a href="#html_out">here</a>)</li>
	  <li>A file named <a href=examples/mcast_example_output_files/cisml.xml><code>cisml.xml</code></a>
	  reporting the matches in XML
	  format using the <a href="http://zlab.bu.edu/~phaverty/SupplementalData/CisML">CisML</a> schema</li>
	  <li>A file named <a href=examples/mcast_example_output_files/mcast.xml><code>mcast.xml</code></a> 
	  describing the inputs to MCAST in XML format and referencing <code>cisml.xml</code></li>
	  <li>A file named <a href=examples/mcast_example_output_files/mcast.txt><code>mcast.txt</code></a>
	  reporting the matches in tab-delimited format
	  (see details <a href="#txt_out"><code>here</code></a>)</li>
	  <li>A file named <a href=examples/mcast_example_output_files/mcast.gff><code>mcast.gff</code></a> 
	  reporting the matches in <a href="http://gmod.org/wiki/GFF3">GFF3 format</a></li>
	</ul>
	<p>The score reported in the GFF3 output is 
	<code>min(1000, -10*(log10(pvalue)))</code>.</p>
      </div>
      <h3>Options</h3>
      <div class="indent">
	<table class="options">
	  <thead>
	    <tr>
	      <th>Option</th>
	      <th>Parameter</th>
	      <th>Description</th>
	      <th>Default Behaviour</th>
	    </tr>
	  </thead>
	  <tbody>
	    <tr>
	      <th colspan="4">General Options</th>
	    </tr>
	    <tr>
	      <td class="popt">--alpha</td><td><span class="pdat">alpha</span></td>
	      <td>
		The fraction of all TF binding sites that are binding sites
		for the TF of interest.
	      </td>
	      <td>1.0</td>
	    </tr>
            <script>
              wropt("all-bg", {"ddash": true, "dna_only": true});
            </script>
	    <tr>
	      <td class="popt">--hardmask</td><td></td>
	      <td>
        Nucleotides in lower case will be converted to the wildcard 'N'.
        This prevents these positions from being considred in motif matches.
        This is useful when the input sequence file has been soft-masked
        for tandem repeats.
        Without hard masking, MCAST may assign sequence segments containing tandem
        repeats a highly significant score.
        </td>
	      <td>Nucleotides in lower case are converted to upper case.</td>
	    </tr>
	    <tr>
	      <td class="popt">--max-gap</td><td><span class="pdat">max gap</span></td>
	      <td>The value of <span class="pdat">max gap</span> specifies the
	      longest distance allowed between two hits in a match. Hits
	      separated by more than <span class="pdat">max gap</span> will be
	      placed in different matches.
	      <b>Note:</b> Large values of <span class="pdat">max gap</span>
	      combined with large values of <span class="pdat">pthresh</span>
	      may prevent MCAST from computing <i>E</i>-values.</td>
	      <td>The maximum gap is set to 50.</td>
	    </tr>
	    <tr>
	      <td class="popt">--max-stored-scores</td><td><span class="pdat">max</span></td>
	      <td>Set the maximum number of scores that will be stored.
	      Keeping a complete list of scores may exceed available memory.
	      Once the number of stored scores reaches the maximum allowed,
	      the least significant 50% of scores will be dropped.
	      In this case, the list of reported motifs may be incomplete and 
	      the q-value calculation will be approximate.</td>
	      <td>The maximum number of stored matches is 100,000.</td>
	    </tr>
	    <tr>
	      <td class="popt">--motif-pthresh</td><td><span class="pdat">pthresh</span></td>
	      <td>sets the scale for calculating pscores for motif hits. The
	      p-score for a hit with p-value <span class="pdat">p</span> is
	      <div class="indent">
		S = -log<sub>2</sub>(<span class="pdat">p</span>/<span class="pdat">pthresh</span>),
	      </div>
	      </td>
	      <td>The motif scaling pvalue defaults to 0.0005.</td>
	    </tr>
	    <script>
	      wropt("all-o", {"ddash": true, "dir": "mcast_out"});
	      wropt("all-oc", {"ddash": true, "dir": "mcast_out"});
	    </script>
	    <tr>
	      <td class="popt">--output-ethresh</td><td><span class="pdat">out <i>E</i>-value</span></td>
	      <td>The <i>E</i>-value threshold for displaying search results. If
	      the <i>E</I>-value of a match is greater than this value, then the
	      match will not be printed. If any combination of
	      <span class="popt">--output-ethresh</span>,
	      <span class="popt">--output-pthresh</span>, or
	      <span class="popt">--output-qthresh</span> is given, whichever
	      option occurs last on the command-line will determine the
	      effective output filter. </td>
	      <td>The <i>E</i>-value threshold is 10.0.</td>
	    </tr>
	    <tr>
	      <td class="popt">--output-pthresh</td><td><span class="pdat">out <i>p</i>-value</span></td>
	      <td>The <i>p</i>-value threshold for displaying search results. If
	      the <i>p</i>-value of a match is greater than this value, then the
	      match will not be printed. If any combination of
	      <span class="popt">--output-ethresh</span>,
	      <span class="popt">--output-pthresh</span>, or
	      <span class="popt">--output-qthresh</span> is given, whichever
	      option occurs last on the command-line will determine the
	      effective output filter.</td>
	      <td>The <i>E</i>-value is used as the threshold. See 
	      <span class="popt">--output-ethresh</span> option.</td>
	    </tr>
	    <tr>
	      <td class="popt">--output-qthresh</td><td><span class="pdat">out <i>q</i>-value</span></td>
	      <td>The <i>q</i>-value threshold for displaying search results. If
	      the <i>q</i>-value of a match is greater than this value, then the
	      match will not be printed. If any combination of
	      <span class="popt">--output-ethresh</span>,
	      <span class="popt">--output-pthresh</span>, or
	      <span class="popt">--output-qthresh</span> is given, whichever
	      option occurs last on the command-line will determine the
	      effective output filter.</td>
	      <td>The <i>E</i>-value is used as the threshold. See 
	      <span class="popt">--output-ethresh</span> option.</td>
	    </tr>
	    <tr>
	      <td class="popt">--parse-genomic-coord</td><td>&nbsp;</td>
	      <td>When this options is specified each sequence header will be
	      checked for UCSC style genomic coordinates. These are of the form:
	      <div style="margin: 5px 0">
		&gt;<span class="pdat">sequence name</span>:<span class="pdat"
		>starting position</span>-<span class="pdat">ending position</span>
	      </div>
	      Where
	      <ul style="margin-top: 0; margin-bottom: 5px">
		<li><span class="pdat">sequence name</span> is the name of the sequence,</li>
		<li><span class="pdat">starting position</span> is the index of the first base and</li>
		<li><span class="pdat">ending position</span> is the index of the final base.</li>
	      </ul>
	      The <span class="pdat">sequence name</span> may not contain any white space.
	      If genomic coordinates are found they will be used as the
	      coordinates in the output. When no coordinates are found the default
	      behaviour is used.</td>
	      <td>The first position in the sequence will be assumed to be 1.</td>
	    </tr>
	    <tr>
	      <td class="popt">--psp</td><td><span class="popt">file</span></td>
	      <td>File containing position-specific priors (PSP) in
	      <a href="meme.html#MEME-PSP">MEME PSP format</a>
	      or 
	      <a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">wiggle format</a>.
	      This file can be generated 
	      using the <a href="create-priors.html">create-priors</a> utility.</td>
	      </td>
	      <td>A uniform position-specific prior is used.</td>
	    </tr>
	    <tr>
	      <td class="popt">--prior-dist</td><td><span class="popt">file</span></td>
	      <td>File containing binned distribution of priors.
	      This file can be generated 
	      using the <a href="create-priors.html">create-priors</a> utility.</td>
	      <td>A uniform position-specific prior is used.</td>
	    </tr>
	    <tr>
	      <td class="popt">--synth</td><td></td>
	      <td>Use synthetic scores for distribution.
        A 0th-order Markov model of nucleotide frequencies will be created by
        choosing a GC content at random between the observed minimum and maximum values.
        This model will be used to generate synthetic sequences, and the synthetic sequences
        will be used to estimate the distribution of p-values.
        </td>
	      <td>No synthetic sequences will be generated.</td>
	    </tr>
	    <tr>
	      <td class="popt">--text</td><td><span class="popt"></span></td>
	      <td>Limits output to plain text sent to standard out.</td>
	      <td></td>
	    </tr>
	    <tr>
	      <td class="popt">--transfac</td><td><span class="popt"></span></td>
	      <td>MCAST will assume that the motif file is in 
        <a href="transfac-format.html">TRANSFAC matrix format</a>.
	      <td>MCAST assumes the motif file is in
        <a href="meme-format.html">MEME format</a>.
         </td>
	    </tr>
            <script>wropt("all-verbosity", {"ddash": true});</script>
            <tr>
              <td class="popt">--version</td><td>&nbsp;</td>
              <td>Display the version and exit.</td>
              <td>Run as normal.</td>
            </tr>
	  </tbody>
	</table>
      </div>

      <h3 id="html_out">HTML output</h3>
      <div class="indent">
	<p style="margin-bottom:0">The HTML output contains</p>
	<ul>
	  <li>A list of the motifs, and the best possible "hit" for each</li>
	  <li>A list of matches. Each match record contains
	  <ul>
	    <li>The name of the sequence</li> 
	    <li>The start position of the motif occurrence (closed, 1-based
	  coordinates, unless genomic coordinates are provided)</li>
	    <li>The end position of the motif occurrence (closed, 1-based
	  coordinates, unless genomic coordinates are provided).</li>
	    <li>The match score</li>
	    <li>The match p-value</li>
	    <li>The match e-value</li>
	    <li>The match q-value</li>
	    <li>A block diagram showing the relative positions of each motif
	    "hit" in the match</li>
	    <li>A detailed view of each match is available as a pop-up window.
	    The detailed view shows the full sequence for the match,
	    the alignment to the motifs, and the <i>p</i>-values for the motif
	    hits</li>
	  </ul></li>
	  <li>The inputs to MCAST</li>
	  <li>Text describing the MCAST results</li>
	</ul>
      </div>

      <h3 id="txt_out">Text output</h3>
      <div class="indent">
	<p style="margin-bottom:0">The plain text output contains a line for
	each match. Each line contains the following fields:</p>
	<ul style="margin-bottom:5px">
	  <li>An id string for the match</li>
	  <li>The name of the sequence containing the match</li>
	  <li>The start position of the motif occurrence (closed, 1-based
	  coordinates, unless genomic coordinates are provided)</li>
	  <li>The end position of the motif occurrence (closed, 1-based
	  coordinates, unless genomic coordinates are provided).</li>
	  <li>The match score</li>
	  <li>The match p-value</li>
	  <li>The match e-value</li>
	  <li>The match q-value</li>
	  <li>The sequence of the matched region</li>
	</ul>
	<p style="margin-top:0">The lines are sorted by score in descending 
	order.</p>
      </div>
    </div>

    <h3>Citing</h3>
    <div class="indent">
      <p>
	If you use MCAST in your research please cite the following paper:<br />
	<span class="citation">
	  Timothy Bailey and William Stafford Noble, 
	  &quot;Searching for statistically significant regulatory modules&quot;, 
	  <i>Bioinformatics (Proceedings of the European Conference on Computational Biology)</i>,
	  19(Suppl. 2):ii16-ii25, 2003.
	  <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">[full text]</a>
	</span>
      </p>
    </div>

    <script src="js/manual.js"></script>

  </body>
</html>
