<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MEME - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
    <style>
      div.example {
        font-family: monospace;
      }
    </style>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/meme_icon.png" alt="MEME Logo">
      <h1>MEME</h1>
      <h2>Multiple Em for Motif Elicitation</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>meme &lt;sequence file&gt; [options]</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <span class="man blurb" data-id="meme"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <div class="indent">
        <br>
        See these <a href="#examples">examples</a> of how
        to use MEME.</p>
      </div>

      <h3>Input</h3>
      <div class="indent">
        <span class='popt'><b>&lt;sequence file&gt;</b></span>
        <div class="indent">
          <p>The name of a file containing <a href="fasta-format.html">FASTA sequences</a>
          or the word <span class="pdat">stdin</span> to indicate that the sequences should be read from
          standard input. Note that MEME does not attempt to detect the alphabet
          from the sequences so you should specify it with the
          <span class="popt">-dna</span>, <span class="popt">-rna</span>,
          <span class="popt">-protein</span> or <span class="popt">-alph</span>
          options.
          MEME also supports a modification to the FASTA format for
          <a href="fasta-format.html#weights">weighting the sequences</a>. Large
          datasets will need to be <a href="fasta-subsample.html">sub-sampled</a>
          to allow MEME to run in <a href="#running_time">reasonable times</a>.</p>
        </div>
      </div>

      <h3>Output</h3>
      <div class="indent">
        <p>MEME outputs its results primarily as an HTML file named
        <a href=examples/meme_example_output_files/meme.html><code>meme.html</code></a>. 
	MEME also outputs a machine-readable XML file and a plain-text versions of its results, 
        named <a href=examples/meme_example_output_files/meme.xml><code>meme.xml</code></a>
        and <a href=examples/meme_example_output_files/meme.txt><code>meme.txt</code></a>, 
	respectively.</p>
        
        <p>These files are placed in a directory named <code>meme_out</code>.
        You can select for the directory to have a different name using the
        <span class="popt">--o</span> or <span class="popt">--oc</span> 
        options.</p>

        <p>MEME gives each motif it discovers a <b>name</b>, which is a 
		<a href=motif-consensus.html>consensus sequence</a>
		that <i>approximately</i> describes the motif.  Motif names are intended
		to be mnemonic only, and are not intended to be used for searching
		sequences for matches; use the position-specific probability matrix (PSPM) 
		contained in each of the MEME output file formats for that purpose 
		via programs such as FIMO and MAST.  
         </p>

      </div>

      <h3>Options</h3>
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th>
              <th>Parameter</th>
              <th>Description</th>
              <th>Default Behaviour</th>
            </tr>
          </thead>

          <tbody>
            <tr>
              <th colspan="4">Output</th>
            </tr>
            <script>
              wropt("all-o", {"ddash": false, "dir": "meme_out"});
              wropt("all-oc", {"ddash": false, "dir": "meme_out"});
            </script>
            <tr>
              <td class="popt">-text</td><td></td>
              <td>Output in text format only to standard output.</td>
              <td>The program behaves as if <code>-oc&nbsp;meme_out</code> had
              been specified.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Alphabet</th>
            </tr>
            <tr>
              <td class="popt">-dna</td><td></td>
              <td>MEME will expect that the sequences use the DNA alphabet. This
                means that the sequence may contain the letters
                "<code>ACGT</code>" and the ambiguous letters
                "<code>BDHKMNRSVWY*-</code>". All ambiguous characters will be
                treated as unknown.
              </td>
              <td>MEME assumes that the sequences are protein.</td>
            </tr>
            <tr>
              <td class="popt">-rna</td><td></td>
              <td>MEME will expect that the sequences use the RNA alphabet. This
                means that the sequence may contain the letters
                "<code>ACGU</code>" and the ambiguous letters
                "<code>BDHKMNRSVWY*-</code>". All ambiguous characters will be
                treated as unknown.
              </td>
              <td>MEME assumes that the sequences are protein.</td>
            </tr>
            <tr>
              <td class="popt">-protein</td><td></td>
              <td>MEME will expect that the sequences use the protein alphabet.
                This means that the sequence may contain the letters
                "<code>ACDEFGHIKLMNPQRSTVWY</code>" and the ambiguous letters
                "<code>BUXZ*-</code>". All ambiguous letters will be treated as
                unknown.</td>
              <td>MEME assumes that the sequences are protein.</td>
            </tr>
            <script>
              wropt("all-alph", {"ddash": false, "meme": true});
            </script>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Contributing Site Distribution</th>
            </tr>
            <tr>
              <td><span class="popt">-mod</span></td><td><span class="pdat">oops</span>|&#8203;<span
                  class="pdat">zoops</span>|&#8203;<span class="pdat">anr</span>|&#8203;<span class="pdat">tcm</span> </td>
              <td>This option is used to describe the distribution of motif sites.
                <table class="dark">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="pdat">oops</span></td><td>One Occurrence Per Sequence</td>
                    <td>MEME assumes that each sequence in the dataset contains
                      exactly one occurrence of each motif. This option is the
                      fastest and most sensitive but the motifs returned by MEME
                      may be "blurry" if any of the sequences is missing 
                      them.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">zoops</span></td><td>Zero or One Occurrence Per Sequence</td>
                    <td>MEME assumes that each sequence may contain at most one
                      occurrence of each motif. This option is useful when you
                      suspect that some motifs may be missing from some of the
                      sequences. In that case, the motifs found will be more
                      accurate than using the first option. This option takes
                      more computer time than the first option (about twice as
                      much) and is slightly less sensitive to weak motifs
                      present in all of the sequences.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">anr</span></td><td>Any Number of Repetitions</td>
                    <td>MEME assumes each sequence may contain any number of
                      non-overlapping occurrences of each motif. This option is
                      useful when you suspect that motifs repeat multiple times
                      within a single sequence. In that case, the motifs found
                      will be much more accurate than using one of the other
                      options. This option can also be used to discover repeats
                      within a single sequence. This option takes the much more
                      computer time than the first option (about ten times as
                      much) and is somewhat less sensitive to weak motifs which
                      do not repeat within a single sequence than the other two
                      options.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">tcm</span></td><td>Any Number of Repetitions</td>
                    <td><span class="pdat">tcm</span> is a synonym for
                    <span class="pdat">anr</span>.</td>
                  </tr>
                </table>
              
              </td>
              <td>MEME assumes the Zero or One Occurrence Per Sequence model.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Number of Motifs</th>
            </tr>
            <tr>
              <td class="popt">-nmotifs</td><td><span class="pdat">n</span></td>
              <td>MEME will stop searching for motifs after finding
                <span class="pdat">n</span> motifs.</td>
              <td>MEME will find 1 motif.</td>
            </tr>
            <tr>
              <td class="popt">-evt</td><td><span class="pdat">ev</span></td>
              <td>MEME will stop searching for motifs if the last motif found
                has an <i>E</i>-value &gt; <span class="pdat">ev</span>.</td>
              <td>MEME will rely on other limits to decide when to stop
                searching for motifs.</td>
            </tr>
            <tr>
              <td class="popt">-time</td><td><span class="pdat">t</span></td>
              <td>MEME will stop searching for motifs if it has found at least
                1 motif and it estimates that finding any more will cause the
                total running time to exceed <span class="pdat">t</span> CPU seconds.</td>
              <td>MEME will rely on other limits to decide when to stop
                searching for motifs.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Number of Motif Occurrences</th>
            </tr>
            <tr>
              <td class="popt">-nsites</td><td><span class="pdat">n</span></td>
              <td>When the motif site distribution model allows, MEME will only
                attempt to find motifs with <span class="pdat">n</span> sites.
                Specifying this option is equalivent to setting the
                <span class="popt">-minsites</span> and <span class="popt">-maxsites</span>
                options to the same value.
                When the distribution model is OOPS this is ignored and the
                number of sites is set to the number of sequences.</td>
              <td>See the <span class="popt">-minsites</span> and
              <span class="popt">-maxsites</span> options for information on the
              default behaviour.</td>
            </tr>
            <tr>
              <td class="popt">-minsites</td><td><span class="pdat">n</span></td>
              <td>When the motif site distribution model allows, MEME will attempt
                to find motifs with at least <span class="pdat">n</span> sites.
                When the distribution model is OOPS this is ignored and the
                number of sites is set to the number of sequences.</td>
              <td>The minimum number of sites is set to 2 when it is not
                otherwise defined by the use of the OOPS model or the
                <span class="popt">-nsites</span> option.</td>
            </tr>
            <tr>
              <td class="popt">-maxsites</td><td><span class="pdat">n</span></td>
              <td>When the motif site distribution model allows, MEME will attempt
                to find motifs with at most <span class="pdat">n</span> sites.
                When the distribution model is OOPS this is ignored and the
                number of sites is set to the number of sequences.</td>
              <td>When the site distribution model is ZOOPS this is set to the
                number of sequences, however when it is ANR this is set to
                <code style="white-space:nowrap">min(5 &times; 
                  <span class="pdat">sequence count</span>, 600)</code>.</td>
            </tr>
            <tr>
              <td class="popt">-wnsites</td><td><span class="pdat">weight</span></td>
              <td>This controls the strength of the bias towards motifs with
                exactly the expected number of sites as defined by the
                <span class="popt">-nsites</span>, <span class="popt">-minsites</span> and
                <span class="popt">-maxsites</span> options. It is a number in
                the range [0..1). The closer to one it is, the stronger the bias
                towards motifs with exactly the expected number of sites.</td>
              <td>The weighting is set to 0.8 .</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Motif Width</th>
            </tr>
            <tr>
              <td class="popt">-w</td><td><span class="pdat">w</span></td>
              <td>Search for motifs with a width of <span class="pdat">w</span>.</td>
              <td>Search for motifs with widths between the range set by
              <span class="popt">-minw</span> and <span class="popt">-maxw</span>.</td>
            </tr>
            <tr>
              <td class="popt">-minw</td><td><span class="pdat">min w</span></td>
              <td>Search for motifs with a width &ge; <span class="pdat">min w</span>.
                </td>
              <td>Searches for motifs with a minimum width of 8.</td>
            </tr>
            <tr>
              <td class="popt">-maxw</td><td><span class="pdat">max w</span></td>
              <td>Search for motifs with a width &le; <span class="pdat">max w</span>.
                </td>
              <td>Searches for motifs with a maximum width of 50.</td>
            </tr>
            <tr>
              <td class="popt">-nomatrim</td><td></td>
              <td>Do not <a href="#multiple_alignment_trimming">adjust motif
                  width</a> using multiple alignments.</td>
              <td>The motif is trimmed to avoid insertions and deletions.</td>
            </tr>
            <tr>
              <td class="popt">-wg</td><td><span class="pdat">wg</span></td>
              <td>The gap opening cost for creating the alignments
                used to 
                <a href="#multiple_alignment_trimming">trim the motif</a>.</td>
              <td>The opening cost for a gap is 11.</td>
            </tr>
            <tr>
              <td class="popt">-ws</td><td><span class="pdat">ws</span></td>
              <td>The gap extension cost for creating the 
                alignments used to 
                <a href="#multiple_alignment_trimming">trim the motif</a>.</td>
              <td>The extension cost for a gap is 1.</td>
            </tr>
            <tr>
              <td class="popt">-noendgaps</td><td></td>
              <td>Do not count end gaps in the alignments used to
                <a href="#multiple_alignment_trimming">trim the motif</a>.</td>
              <td>End gaps are penalised like any other gap.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Background Model</th>
            </tr>
            <tr>
              <td class="popt">-bfile</td><td><span class="pdat">bfile</span></td>
              <td>The name of <a href="bfile-format.html">Markov background model</a> file.  
		You can use a background model with MEME in order 
	      	to normalize for biased distribution of letters and groups of letters in your sequences.
      		A 0-order model adjusts for single letter biases, a 1-order model adjusts for
      		dimer biases (e.g., GC content in DNA sequences), etc.
                <p style="margin-bottom:0">The background model is used by MEME:</p>
                  <ol style="margin-top:0">
                    <li>during EM as the "null mode",</li>
                    <li>for calculating the log liklihood ratio of a motif,</li>
                    <li>for calculating the significance (<i>E</i>-value) of a motif, and,</li>
                    <li>for creating the position-specific scoring matrix (log-odds matrix).</li>
                  </ol>
                  <p>Note that MEME uses only the 0-order portion (single letter
                  frequencies) of the background model for purposes 3 and 4,
                  but uses the full-order model for purposes 1 and 2, above.</p>
              </td>
              <td>The 0-order background frequencies are determined from the
                sequences.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Position-Specific Priors</th>
            </tr>
            <tr>
              <td class="popt">-psp</td><td><span class="pdat">PSP file</span></td>
              <td>The name of a <a href="psp-format.html">MEME position specific
                  priors</a> file. This can be used to bias the search for
                motifs in MEME. The <span class="pdat">PSP file</span> supplies
                a position-specific prior distribution on the location of motif
                sites in sequence(s) in the input dataset.</td>
              <td>All motif sites are considered equally likely.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Palindromes &amp; Strands</th>
            </tr>
            <tr>
              <td class="popt">-revcomp</td><td></td>
              <td>Consider both the given strand and the reverse complement
                strand when searching for motifs in a complementable alphabet (ie DNA).</td>
              <td>Search for motifs of complementable alphabets on the given strand only.</td>
            </tr>
            <tr>
              <td class="popt">-pal</td><td></td>
              <td>This causes MEME to only look for palindromes in complementable datasets.
                MEME averages the letter frequencies in complementary columns of
                the motif (PSPM) together. For instance, if the width of the
                motif is 10, columns 1 and 10, 2 and 9, 3 and 8, etc., are
                averaged together. For DNA the averaging would combines the
                frequency of A in one column with T in the other,
                and the frequency of C in one column with G in the other.</td>
              <td>MEME won't specifically look for palindromes.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">EM Algorithm</th>
            </tr>
            <tr>
              <td class="popt">-maxiter</td><td><span class="pdat">max iter</span></td>
              <td>The number of iterations of EM to run from any starting point.
                EM is run for <span class="pdat">max iter</span> iterations or
                until convergence (see <span class="popt">-distance</span>,
                below) from each starting point.</td>
              <td>MEME will use a maximum of 50 EM iterations from any starting point.</td>
            </tr>
            <tr>
              <td class="popt">-distance</td><td><span class="pdat">distance</span></td>
              <td>MEME stops iterating EM when the change in the motif frequency
                matrix is less than <span class="pdat">distance</span>. Change is
                defined as the euclidean distance between two successive
                frequency matrices.</td>
              <td>The distance used for measuring convergence is 0.001.</td>
            </tr>
            <tr>
              <td class="popt">-prior</td><td><span class="pdat">dirichlet</span>|&#8203;<span 
                  class="pdat">dmix</span>|&#8203;<span class="pdat">mega</span>|&#8203;<span 
                  class="pdat">megap</span>|&#8203;<span class="pdat">addone</span></td>
              <td>The type of prior to use.
                <table class="light">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="pdat">dirichlet</span></td><td>Simple Dirichlet</td>
                    <td>This is the default for DNA sequences.
                      It is a simple dirichlet prior based on the background
                      model as set by the <span class="popt">-bfile</span>
                      option.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">dmix</span></td><td>Dirichlets Mix</td>
                    <td>This is the default for protein sequences when using the
                      OOPS distribution model. It is a mixture of Dirichlets
                      prior. The source of the Dirichlets is specified by the
                      <span class="popt">-plib</span> option.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">mega</span></td><td>Mega-weight Dirichlets Mix</td>
                    <td>This is an extremely low variance version of
                      <span class="popt">dmix</span>. The variance is scaled
                      inversely with the size of the dataset.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">megap</span></td><td>Mega-weight Dirichlets Mix Plus</td>
                    <td>This is the default for protein sequences when using the
                      ZOOPS or ANR models. This behaves like
                      <span class="popt">mega</span> until the last iteration of
                      EM when it reverts to <span class="popt">dmix</span>
                      behaviour.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">addone</span></td><td>Add One</td>
                    <td>Add +1 to each observed count.</td>
                  </tr>
                </table>
              </td>
              <td>
                When the prior is not selected:
                <table class="light" style="width:100%">
                  <tr>
                    <th>Model</th><th>DNA</th><th>Protein</th>
                  </tr>
                  <tr>
                    <td>One Occurrence Per Sequence</td>
                    <td><span class="pdat">dirichlet</span></td>
                    <td><span class="pdat">dmix</span></td>
                  </tr>
                  <tr>
                    <td>Zero or One Occurrence Per Sequence</td>
                    <td><span class="pdat">dirichlet</span></td>
                    <td><span class="pdat">megap</span></td>
                  </tr>
                  <tr>
                    <td>Any Number of Repetitions</td>
                    <td><span class="pdat">dirichlet</span></td>
                    <td><span class="pdat">megap</span></td>
                  </tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-b</td><td><span class="pdat">b</span></td>
              <td>The strength of the prior on model parameters: 
                <span class="pdat">b</span> = 0 means use intrinsic strength of
                prior for prior = <span class="pdat">dmix</span>.</td>
              <td>
                <table class="light" style="width:100%">
                  <tr>
                    <th>Prior</th><th><span class="pdat">b</span></th>
                  </tr>
                  <tr><td><span class="pdat">dirichlet</span></td><td>0.01</td></tr>
                  <tr><td><span class="pdat">dmix</span></td><td>0</td></tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-plib</td><td><span class="pdat">plib</span></td>
              <td>The name of the file containing the
                <a href="dmix-format.html">Dirichlet mixtures prior library</a>.</td>
              <td>
                The default value of <span class="pdat">plib</span> depends on
                the alphabet of the sequences.
                <table class="light" style="width:100%">
                  <tr>
                    <th>Alphabet</th><th><span class="pdat">plib</span></th>
                  </tr>
                  <tr><td>DNA</td><td><code>INSTALL_DIR/etc/dna.plib</code></td></tr>
                  <tr><td>Protein</td><td><code>INSTALL_DIR/etc/prior30.plib</code></td></tr>
                </table>
              </td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Selecting Starts for EM</th>
            </tr>
            <tr>
              <td class="popt">-spfuzz</td><td><span class="pdat">fuzz</span></td>
              <td>The fuzziness of sequence to theta mapping. The meaning of this
                parameter depends on the choice of mapping function as set by the
                <span class="popt">-spmap</span> option.</td>
              <td>
                The default value of <span class="pdat">fuzz</span> depends on
                the mapping function (see <span class="popt">-spmap</span>).
                <table class="dark" style="width:100%">
                  <tr>
                    <th>Mapping Function</th><th><span class="pdat">fuzz</span></th>
                  </tr>
                  <tr>
                    <td>Uniform</td><td>0.5</td>
                  </tr>
                  <tr>
                    <td>Point Accepted Mutation</td><td>120</td>
                  </tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-spmap</td><td><span class="pdat">uni</span>|<span class="pdat">pam</span></td>
              <td>The mapping function to use for estimating theta.
                <table class="dark">
                  <tr>
                    <th>Value</th><th>Name</th><th>Description</th>
                  </tr>
                  <tr>
                    <td><span class="pdat">uni</span></td><td>Uniform</td>
                    <td>Add a uniform prior of <span class="pdat">fuzz</span>
                    when converting a substring to an estimate of theta.</td>
                  </tr>
                  <tr>
                    <td><span class="pdat">pam</span></td><td>Point Accepted Mutation</td>
                    <td>Use <a href="http://en.wikipedia.org/wiki/Point_accepted_mutation">PAM</a>
                      matrices, with the number of mutation events set by
                      <span class="pdat">fuzz</span>, to estimate theta.</td>
                  </tr>
                </table>
              </td>
              <td>
                The default mapping function depends on the alphabet of the sequences.
                <table class="dark" style="width:100%">
                  <tr>
                    <th>Alphabet</th><th>Mapping Function</th>
                  </tr>
                  <tr>
                    <td>DNA</td><td>Uniform</td>
                  </tr>
                  <tr>
                    <td>Protein</td><td>Point Accepted Mutation</td>
                  </tr>
                </table>
              </td>
            </tr>
            <tr>
              <td class="popt">-cons</td><td><span class="pdat">consensus</span></td>
              <td>Override the sampling of starting points and just use a
                starting point derived from <span class="pdat">consensus</span>.
                This is useful when an actual occurrence of a motif is known and
                can be used as the starting point for finding the motif.</td>
              <td>Refer to the <span class="popt">-spmap</span> option for the
              default behaviour.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Branching Search on EM Starts</th>
            </tr>
            <tr>
              <td class="popt">-heapsize</td><td><span class="pdat">hs</span></td>
              <td>Size of heaps for widths where substring search occurs.
              See the <a href="#branching_search">branching search</a> section
              for more details.</td>
              <td></td>
            </tr>
            <tr>
              <td class="popt">-x_branch</td><td></td>
              <td><span class="experimental">Experimental</span>
                Perform x-branching. This does a <a 
                  href="http://en.wikipedia.org/wiki/Beam_search">BEAM search</a>
                and tests the actual words in the input as well as words at
                hamming distance 1 at each successsive level of branching.</td>
              <td>Normal branching search is performed. See the
              <a href="#branching_search">branching search</a> section for more
              details.</td>
            </tr>
            <tr>
              <td class="popt">-w_branch</td><td></td>
              <td><span class="experimental">Experimental</span>
                Perform width branching. This is not optimised.</td>
              <td>Normal branching search is performed. See the 
              <a href="#branching_search">branching search</a> section for more
              details.</td>
            </tr>
            <tr>
              <td class="popt">-bfactor</td><td><span class="pdat">bf</span></td>
              <td>The number of iterations of branching search. See the
              <a href="#branching_search">branching search</a> section for more
              details.</td>
              <td>MEME uses 3 iterations of branching search.</td>
            </tr>
          </tbody>

          <tbody>
            <tr>
              <th colspan="4">Misc</th>
            </tr>
            <tr>
              <td class="popt">-maxsize</td><td><span class="pdat">max size</span></td>
              <td>Maximum allowed dataset size (in characters).</td>
              <td>Maximum allowed dataset size 100000 characters.  Note that the default maximum
                size exists to warn users that their dataset is possibly too
                large to process in a reasonable time so please consider
                carefully before increasing this value.</td>
            </tr>
            <tr>
              <td class="popt">-nostatus</td><td></td>
              <td>Print no status messages to terminal.</td>
              <td>Print minimal status messages to terminal.</td>
            </tr>
            <tr>
              <td class="popt">-p</td><td><span class="pdat">np</span></td>
              <td>Use faster, parallel version of MEME with <span class="pdat">np</span> processors.
                  The parameter <span class="pdat">np</span> may be a number or it 
                  may be a quoted string <b>starting</b> with a number and followed by arguments
                  to the particular MPI run command for your installation (e.g., <code>mpirun</code>).</td>
              <td>Use a single processor.</td>
            </tr>
            <tr>
              <td class="popt">-sf</td><td><span class="pdat">sf</span></td>
              <td>Print <span class="pdat">sf</span> as name of sequence file.</td>
              <td>Print actual file name.</td>
            </tr>
            <tr>
              <td class="popt">-V</td><td></td>
              <td>Print extensive status messages to terminal.</td>
              <td>Print minimal status messages to terminal.</td>
            </tr>
            <tr>
              <td class="popt">-h</td><td></td>
              <td>Display a usage message and exit.</td>
              <td>Run as normal.</td>
            </tr>
            <tr>
              <td class="popt">-version</td><td></td>
              <td>Display the version and exit.</td>
              <td>Run as normal.</td>
            </tr>
          </tbody>

        </table>
      </div>

      <h3 id="objective_function">Motif Objective Function</h3>
      <div class="indent">
        <p>MEME uses an objective function on motifs to select the "best" motif.
        The objective function is based on the statistical significance of the
        log likelihood ratio (LLR) of the occurrences of the motif. The
        <i>E</i>-value of the motif is an estimate of the number of motifs (with
        the same width and number of occurrences) that would have equal or
        higher log likelihood ratio if the input sequences had been
        generated randomly according to the (0-order portion of the) background
        model.</p>

        <p>MEME searches for the motif with the smallest <i>E</i>-value. It
        searches over different motif widths, numbers of occurrences, and
        positions in the input sequences for the motif occurrences. The user may
        limit the range of motif widths and number of occurrences that MEME
        tries. In addition, MEME trims the motif (using a dynamic programming
        multiple alignment) to eliminate any positions where there is a gap in
        any of the occurrences.</p>

        <p>The log likelihood ratio of a motif is
        <code>llr = log (Pr(sites | motif) / Pr(sites | back))</code> and is a
        measure of how different the sites are from the background model.
        <code>Pr(sites | motif)</code> is the probability of the occurrences
        given the a model consisting of the position-specific probability matrix
        (PSPM) of the motif. <code>Pr(sites | back)</code> is the probability
        of the occurrences given the background model. The background model is
        an <a href="bfile-format.html">n-order Markov model</a>. By default,
        it is a 0-order model consisting of the frequencies of the letters in
        the input sequences. A different 0-order Markov model or higher order
        Markov models can be specified to MEME using the
        <span class="popt">-bfile</span> option.</p>

        <p>The <i>E</i>-value reported by MEME is actually an approximation of
        the <i>E</i>-value of the log likelihood ratio (an approximation is used
        because it is far more efficient to compute). The approximation is based
        on the fact that the log likelihood ratio of a motif is the sum of the
        log likelihood ratios of each column of the motif. Instead of computing
        the statistical significance of this sum (its <i>p</i>-value), MEME
        computes the <i>p</i>-value of each column and then computes the
        significance of their product. Although not identical to the
        significance of the log likelihood ratio, this easier to compute
        objective function works very similarly in practice.</p>

        <p>The motif significance is reported as the <i>E</i>-value of the
        motif. The statistical significance of a motif is computed based on:</p>
        <ol>
          <li>the log likelihood ratio,</li>
          <li>the width of the motif,</li>
          <li>the number of occurrences,</li>
          <li>the 0-order portion of the background model,</li>
          <li>the size of the input sequences, and</li>
          <li>the type of model (<span class="pdat">oops</span>,
          <span class="pdat">zoops</span>, or
          <span class="pdat">anr</span>, which determines the
          number of possible different motifs of the given width and number of
          occurrences).</li>
        </ol>

        <p>MEME searches for motifs by performing Expectation Maximization (EM)
        on a motif model of a fixed width and using an initial estimate of the
        number of sites. It then sorts the possible sites according to their
        probability according to EM. MEME then calculates the <i>E</i>-values of
        the first n sites in the sorted list for different values of n. This
        procedure (first EM, followed by computing <i>E</i>-values for different
        numbers of sites) is repeated with different widths and different
        initial estimates of the number of sites. MEME outputs the motif with
        the lowest <i>E</i>-value.</p>

      </div>

      <h3 id="multiple_alignment_trimming">Multiple Alignment Trimming</h3>
      <div class="indent">
        <p>Once a candidate motif has been found the multiple alignment method
        performs a separate pairwise alignment of the site with the highest
        probability and each other possible site (the alignment includes
        width/2 positions on either side of the sites).</p>

        <p>The pairwise alignments are then combined and the method determines
        the widest section of the motif with no insertions or deletions. If this
        alignment is shorter than <span class="pdat">min w</span>, it tries to
        find an alignment allowing up to one insertion/deletion per motif column.
        This continues (allowing up to 2, 3 ... insertions/deletions per motif
        column) until an alignment of width at least <span class="pdat">min
          w</span> is found.</p>

        <p>The switches <span class="popt">-nomatrim</span>, 
        <span class="popt">-wg</span>, <span class="popt">-ws</span> and
        <span class="popt">-noendgaps</span> control trimming of
        motifs using the multiple alignment method. Specifying the
        <span class="popt">-nomatrim</span> option causes MEME to skip trimming
        altogether. The <span class="popt">-wg</span> and
        <span class="popt">-ws</span> options set the costs of including gaps in
        the alignments and the <span class="popt">-noendgaps</span> allows gaps
        at the beginning and end to be treated specially.</p>
        
        <p>After trimming, the number of occurrences is then adjusted to
        maximize the motif <i>E</i>-value, and then the motif width is further
        shortened to optimize the <i>E</i>-value.</p>

      </div>

      <h3 id="branching_search">Branching Search</h3>
      <div class="indent">
        <p>The search for good EM starting points can be improved by using
        branching search.</p>

        <p>Branching search begins with a fixed-sized heap of best EM starts
        identified during the search of subsequences from the dataset. These
        starts are also called "seeds". The fixed-sized heap of seeds is used as
        the "branch_heap" during the first iteration of branching search.</p>

        <p>For each iteration of branching search, all seeds in the current
        branch_heap are considered. All seeds in the ball within hamming
        distance 1 of a given seed are evaluated and added to a new heap. The
        ball of new seeds is generated by mutating each character of the initial
        seed to each alternative character in the alphabet.</p>

        <p>After the ball for every branch_heap seed has been evaluated, the
        seeds in the resulting new heap are added to the heap of best EM starts.
        The new heap is then used as the branch_heap for the next iteration of
        branching search.</p>

      </div>

      <h3 id="running_time">Running time on large inputs</h3>
      <div class="indent">
        <p>MEME's run time is <b>cubic</b> with respect to the number of input
        sequences. Datasets with more than 1000 input sequences in OOPS mode
        or 1000+ sites in ZOOPS and ANR mode are intractable in terms
        of run time. For example, on a 3 GHz CPU, initialization of the
        <i>p</i>-value table takes about 10 min for 1000 sites/sequences in OOPS,
        about a day for 5000 sites, and about 1 week for 10,000 sites. Then
        the quadratic run time with respect to the total number of characters
        kicks in next.</p>

        <p>MEME's run time is <b>quadratic</b> with respect to the number of 
        characters. So when the input dataset size doubles, the run time
        quadruples. If a 30,000 character dataset takes 15 min on a single
        cpu, a 30,000,000 character dataset would take 1 million times longer
        (N squared), or roughly 250,000 CPU hours, or 28.5 CPU years.</p>

        <p>The parallel version of MEME scales up to about 128 processors.
        Please see <a href="http://www.sdsc.edu/~tbailey/papers/cabios96.pdf"
        >http://www.sdsc.edu/~tbailey/papers/cabios96.pdf</a> for a
        discussion of the parallel program. You must bear in mind however,
        that doubling the total number of sequences input to MEME means
        that you will need 8 times more processors for the job to finish in
        the same amount of time.</p>
      </div>

      <h3 id="examples">Examples</h3>
      <div class="indent">
        <p>The following examples use data files provided in this release of the
        MEME Suite (in the tests directory).</p>
        <ol>
          <li>
          <p>A simple DNA example:</p>
          <div class="example">meme crp0.s -dna -mod oops -pal</div>
          <p>MEME looks for a single motif in the file <code>crp0.s</code> which
          contains DNA sequences in FASTA format. The OOPS model is used so MEME
          assumes that every sequence contains exactly one occurrence of the
          motif. The palindrome switch is given so the motif model (PSPM) is
          converted into a palindrome by combining corresponding frequency
          columns.  MEME automatically chooses the best width for the motif in
          this example since no width was specified.</p>
          </li>

          <li>
          <p>Searching for motifs on both DNA strands:</p>
          <div class="example">meme crp0.s -dna -mod oops -revcomp</div>
          <p>This is like the previous example except that the
          <span class="popt">-revcomp</span> switch tells MEME to consider both
          DNA strands, and the <span class="popt">-pal</span> switch is absent
          so the palindrome conversion is omitted. When DNA uses both DNA
          strands, motif occurrences on the two strands may not overlap. That
          is, any position in the sequence given in the input sequences may be
          contained in an occurrence of a motif on the positive strand or the
          negative strand, but not both.</p>
          </li>

          <li>
          <p>A fast DNA example:</p>
          <div class="example">meme crp0.s -dna -mod oops -revcomp -w 20</div>
          <p>This example differs from example 1) in that MEME is told to only 
          consider motifs of width 20. This causes MEME to execute about 10
          times faster. The <code>-w</code> switch can also be used with protein
          datasets if the width of the motifs is known in advance.</p>
          </li>

          <li>
          <p>Using a higher-order background model:</p>
          <div class="example">meme INO_up800.s -dna -mod anr -revcomp -bfile yeast.nc.6.freq</div>
          <p>In this example we use <code>-mod anr</code> and <code>-bfile
            yeast.nc.6.freq</code>. This specifies that</p>
          <ol type="a">
            <li>the motif may have any number of occurrences in each sequence, and,
            <li>the Markov model specified in yeast.nc.6.freq is used as the 
            background model. This file contains a fifth-order Markov model for
            the non-coding regions in the yeast genome.</li>
          </ol>
          <p>Using a higher order background model can often result in more
          sensitive detection of motifs. This is because the background model
          more accurately models non-motif sequence, allowing MEME to
          discriminate against it and find the true motifs.</p>
          </li>

          <li>
          <p>A simple protein example:</p>
          <div class="example">meme lipocalin.s -mod oops -maxw 20 -nmotifs 2</div>
          <p>The <code>-dna</code> switch is absent, so MEME assumes the file
          <code>lipocalin.s</code> contains protein sequences. MEME searches for
          two motifs each of width less than or equal to 20. (Specifying
          <code>-maxw 20</code> makes MEME run faster since it does not have to
          consider motifs longer than 20.) Each motif is assumed to occur in
          each of the sequences because the OOPS model is specified.</p>
          </li>

          <li>
          <p>Another simple protein example:</p>
          <div class="example">meme farntrans5.s -mod anr -maxw 40 -maxsites 50</div>
          <p>MEME searches for a motif of width up to 40 with up to 50
          occurrences in the entire set of input sequences. The ANR sequence model is
          specified, which allows each motif to have any number of occurrences
          in each sequence. This dataset contains motifs with multiple repeats
          of motifs in each sequence. This example is fairly time consuming due
          to the fact that the time required to initialize the motif probability
          tables is proportional to <span class="pdat">max width</span> &times;
          <span class="pdat">max sites</code>.</p>
          </li>

          <li>
          <p>A much faster protein example:</p>
          <div class="example">meme farntrans5.s -mod anr -w 10 -maxsites 30 -nmotifs 3</div>
          <p>This time MEME is constrained to search for three motifs of width
          exactly ten. The effect is to break up the long motif found in the
          previous example. The <span class="popt">-w</span> switch forces
          motifs to be <b>exactly</b> ten letters wide. This example is much faster
          because, since only one width is considered, the time to build the
          motif probability tables is only proportional to
          <span class="pdat">max sites</span>.</p>
          </li>

          <li>
          <p>Splitting the sites into three:</p>
          <div class="example">meme farntrans5.s -mod anr -maxw 12 -nsites 24 -nmotifs 3</div>
          <p>This forces each motif to have 24 occurrences, exactly, and be up
          to 12 letters wide.</p>
          </li>

          <li>
          <p>A larger protein example with <i>E</i>-value cutoff:</p>
          <div class="example">meme adh.s -mod zoops -nmotifs 20 -evt 0.01</div>
          <p>In this example, MEME looks for up to 20 motifs, but stops when a
          motif is found with <i>E</i>-value greater than 0.01. Motifs with
          large <i>E</i>-values are likely to be statistical artifacts rather
          than biologically significant.</p>
          </li>

        </ol>
      </div>
    </div>

    <h3 id="citing">Citing</h3>
    <div class=indent>
      If you use MEME in your research, please cite the following paper:<br />
      <div class="citation">
        Timothy L. Bailey and Charles Elkan,
        &quot;Fitting a mixture model by expectation maximization to
        discover motifs in biopolymers&quot;,
        <em>Proceedings of the Second International Conference on Intelligent Systems
        for Molecular Biology</em>, pp. 28-36, AAAI Press, Menlo Park, California, 1994. 
        <a href="http://www.sdsc.edu/~tbailey/papers/ismb94.ps">[postscript]</a> 
        <a href="http://www.sdsc.edu/~tbailey/papers/ismb94.pdf">[pdf]</a>
      </div>
    </div>

    <script src="js/manual.js"></script>

  </body>
</html>

