<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>MoMo - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/momo_icon.png" alt="MOMO Logo">
      <h1>MoMo</h1>
      <h2>Modification Motifs</h2>
    </div>
    
    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>momo &lt;algorithm&gt; [options] &lt;PSM file&gt;</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
       <span class="man blurb" data-id="momo"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <div class="indent">

      </div>

      <h3>Input</h3>
      <div class="indent">

        <h4>Algorithm</h4>
        <p>The algorithm used to search for motifs. Available algorithms include "simple", "motifx", and "modl".
        <ul>
          <li>simple: Creates a PWM based on the frequencies of all the passing modifications and their flanking amino acids.</li>
          <li>motifx: Searches through the list of passing modifications using the algorithm described in the <a href="http://motif-x.med.harvard.edu">Schwartz and Gygi</a> paper.
          <br>Schwartz, D. and Gygi, S. P. (2005). An iterative statistical approach to the identification of protein phosphorylation motifs from large-scale data sets. Nature Biotechnology, 23(11), 1391–1398.</li>
          <li>modl: Searches through the list of passing modifications using the algorithm described in the <a href="http://compbio.cs.brown.edu/software/">Ritz, et. al</a> paper.
          <br>Ritz, A., Shakhnarovich, G., Salomon, A., and Raphael, B. (2009). Discovery of phosphorylation motif mixtures in phosphoproteomics data. Bioinformatics, 25(1), 14–21.</li>
        </ul>
        </p>

        <h4>PSM file</h4>
        <p>A tab-delimited peptide-spectrum mmatch (PSM) file in which each row corresponds to a mass spectrum annotated with its    corresponding peptide sequence.  Currently, only PSM files generated by using the <a href="http://cruxtoolkit.sourceforge.net">Tide search engine</a> are supported.</p>
        

      </div>

      <h3>Output</h3>
      <div class="indent">

        <p>MoMo will create a directory, named <code>momo_out</code> by default.
        Any existing output files in the directory will be overwritten. The
        directory will contain:</p>
        <ul>
          <li>A file (<code>momo.txt</code>) containing the PTM motifs in text format.</li>

          <li>A file (<code>momo.html</code>) containing the same motifs in HTML format.</li>

        <p>The default output directory can be overridden using the
        <span class="popt">--o</span> or <span class="popt">--oc</span> options
        which are described below.</p>

      </div>

      <h3>Options</h3>
      <div class="subcontent">
        <input type="radio" name="algorithm" id="alg_simp" value="alg_simp" checked>
          <label for="alg_simp">Simple</label>
        <input type="radio" name="algorithm" id="alg_mtfx" value="alg_mtfx">
          <label for="alg_mtfx">Motif-X</label>
        <input type="radio" name="algorithm" id="alg_modl" value="alg_modl">
          <label for="alg_modl">MoDL</label>
      </div>

      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th>
              <th>Parameter</th>
              <th>Description</th>
              <th>Default Behaviour</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="4">General Options</th>
            </tr>
            <tr id="bg_filetype">
              <td class="popt">--bg-filetype</td><td><span class="popt">fasta|prealigned</span></td>
              <td>This option inidicates the format of the the protein database.</td>
              <td>The protein database is specified as a fasta file.</td>
            </tr>
            <tr id="count_threshold">
              <td class="popt">--count-threshold</td><td><span class="popt">num</span></td>
              <td>This option inidicates minimum number of sequences in the phosphorylation data set needed to match the residue/position pair for each recursive iteration of motif-x.</td>
              <td>The default count threshold is 20 occurrences.</td>
            </tr>
            <tr>
              <td class="popt">--eliminate-repeats</td><td><span class="popt">num</span></td>
              <td>This option will remove duplicate copies of
                modifications with identical flanking sequences.  The integer parameter specifies the width of the region used to determine identify. Because the window is symmetric around the central, modified amino acid, the width parameter must be odd.  To turn this option off, specify a width of 0.</td>
              <td>All modifications that are identical up to 7 amino acids long (3 on each side of the modification) are removed.</td>
            </tr>
            <tr>
              <td class="popt">--fg-filetype</td><td><span class="popt">fasta|prealigned|psm</span></td>
              <td>This option inidicates the format of the input file with the phosphorylated sequences.</td>
              <td>The input is specified as a psm file.</td>
            </tr>
            <tr>
              <td class="popt">--filter</td><td><span class="popt">[field],lt|le|eq|ge|gt,[threshold]
              </span></td>
              <td>Only PSMs with scores better than the specified
                threshold are accepted for analysis. The "[field]"
                component of the parameter specifies the name of the
                column from which the score is drawn.  The next
                component specifies whether PSMs with scores less
                than, less than or equal, etc. are retained.  The
                third component is the threshold itself.
              </td>
              <td>No filter.</td>
            </tr>
            <tr id="hash_fasta">
              <td class="popt">--hash-fasta</td><td><span class="popt">num</span></td>
              <td>If a protein database is provided, the process of
                finding the location of the peptide within the protein
                can be sped up using an O(1) lookup table hashing from
                each unique kmer to an arraylist of locations. The number
                specified is used as an argument to the kmer length. If the
                number specified is 0, then the program will proceed using
                linear search instead of creating a lookup table.
              </td>
              <td>Create an O(1) lookup table with kmer length 6.</td>
            </tr>
            <tr id="max_iterations">
              <td class="popt">--max-iterations</td><td><span class="popt">num</span></td>
              <td>The maximum number of iterations for MoDL before it stops.
              </td>
              <td>MoDL will stop after 50 iterations.</td>
            </tr>
            <tr id="max_motifs">
              <td class="popt">--max-motifs</td><td><span class="popt">num</span></td>
              <td>The maximum number of motifs MoDL is allowed to find.
              </td>
              <td>MoDL cannot allow more than 100 motifs.</td>
            </tr>
            <tr id="no_stop_decrease_iteration">
              <td class="popt">--no-stop-decrease-iteration</td><td><span class="popt">num</span></td>
              <td>MoDL will stop if there is no decrease in MDL after several iterations.
              </td>
              <td>MoDL stops after 10 iterations of no decrease or equal.</td>
            </tr>
            <tr id="min_occurrences">
              <td class="popt">--min-occurences</td><td><span class="popt">num</span></td>
              <td>A motif will only be constructed if the specified number of
                occurrences is reached. This threshold is applied after eliminating
                repeats.
              </td>
              <td>Only print motif if pattern occurs at least 5 times.</td>
            </tr>
            <script>
              wropt("all-o", {"ddash": true, "dir": "momo_out"});
              wropt("all-oc", {"ddash": true, "dir": "momo_out"});
            </script>
            <tr id="protein_database">
              <td class="popt">--protein-database</td><td><span class="popt">protein database file</span></td>
              <td>The protein database used to generate the PSM file. If provided, this file
                is used to find the amino acids flanking each modification and also to generate
                the background frequencies.</td>
              <td>Flanking sequences are derived from the given PSM
              file, substituting "don't care" symbols for missing
              entries.  Amino acid background frequencies are derived
              from the non-redundant protein database.</td>
            </tr>
            <tr>
              <td class="popt">--remove-unknowns</td><td><span class="popt">T|F</span></td>
              <td>This option inidicates whether to remove all sequences that contain an 'X'.</td>
              <td>Do not remove unknown sequences.</td>
            </tr>
            <tr id="score_threshold">
              <td class="popt">--score-threshold</td><td><span class="popt">num</span></td>
              <td>This option inidicates the largest bionmial probability for a residue/position pair to be counted as significant during each recursive iteration of motif-x.</td>
              <td>The binomial probability must be smaller than 0.000001 to be considered significant.</td>
            </tr>
            <tr>
              <td class="popt">--single-motif-per-mass</td><td><span class="popt">T|F</span></td>
              <td>Provides the option of generating a single motif for each distinct
                modification mass. For example, phosphorylation is typically specified
                as a mass of 79.97 added to the amino acids S, T or Y. If this parameter
                is set to false, then three separate phosphorylation motifs are
                generated, each with a perfectly conserved central amino acid. If true,
                then all the phosphorylation events are combined into a single
                motif, with a mixture of S, T and Y in the central position.</td>
              <td>Default is set to false.</td>
            </tr>
            <script>wropt("all-verbosity", {"ddash": true});</script>
            <tr>
              <td class="popt">--version</td><td>&nbsp;</td>
              <td>Display the version and exit.</td>
              <td>Run as normal.</td>
            </tr>
            <tr>
              <td class="popt">--width</td><td><span class="popt">num</span></td>
              <td>This is an integer specifying the width of the
                motif. Because the motif is symmetric around the
                central, modified amino acid, the width parameter
                should be odd. </td>
              <td>Motifs of width 7 are generated.</td>
            </tr>
            <tr>
          </tbody>
        </table>
      </div>
    </div>
		
    <script>
      
      document.getElementById("alg_simp").addEventListener("click", on_alg_change, false);
      document.getElementById("alg_mtfx").addEventListener("click", on_alg_change, false);
      document.getElementById("alg_modl").addEventListener("click", on_alg_change, false);
      on_alg_change();
      
      function on_alg_change() {
        document.getElementById("count_threshold").style.display = (document.getElementById('alg_mtfx').checked ? 'table-row' : 'none');
        document.getElementById("min_occurrences").style.display = (document.getElementById('alg_simp').checked ? 'table-row' : 'none');
        document.getElementById("score_threshold").style.display = (document.getElementById('alg_mtfx').checked ? 'table-row' : 'none');
        
        document.getElementById("max_iterations").style.display = (document.getElementById('alg_modl').checked ? 'table-row' : 'none');
        document.getElementById("max_motifs").style.display = (document.getElementById('alg_modl').checked ? 'table-row' : 'none');
        document.getElementById("no_decrease_stop_iteration").style.display = (document.getElementById('alg_modl').checked ? 'table-row' : 'none');
        
      }
    </script>

    <h3>Citing</h3>
    <div class="indent">
      <p>If you use MoMo in your research, please cite the following paper:
    </p>
    </div>

    <script src="js/manual.js"></script>

  </body>
</html>
