<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Tomtom - MEME Suite</title>
    <link rel="shortcut icon" type="image/x-icon" href="images/memesuite_icon.ico">
    <link rel="icon" type="image/png" href="images/memesuite_icon.png">
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <script src="js/menu.js"></script>
    <script src="js/shared-doc.js"></script>
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <div class="prog_logo">
      <img src="images/tomtom_icon.png" alt="Tomtom Logo">
      <h1>Tomtom</h1>
      <h2>Motif Comparison Tool</h2>
    </div>

    <div style='display:inline' id='man_web_button'>
      <p> <button onclick="toggle_man_display()">View Web Version</button> </p>
    </div>

    <div style='display:none' id='man_cmd_button'>
      <p> <button onclick="toggle_man_display()">View Command-Line Version</button> </p>
    </div>

    <div style='display:inline' id='man_usage'>
      <div class="usage">
        <h3>Usage:</h3>
        <code>tomtom [options] &lt;query&nbsp;motifs&gt; 
          &lt;target&nbsp;motif&nbsp;database&gt;+</code>
      </div>
    </div>

    <h3>Description</h3>
    <div class="indent">
      <span class="man blurb" data-id="tomtom"></span>
    </div>

    <!-- Command Line Version Doc -->
    <div style='display:inline' id='man_cmd'>
      <h3>Inputs</h3>
      <div class="indent">
        <h4>Query Motifs</h4>
        <div class="indent">
          <p>A file containing one or more motifs in <a href="meme-format.html"
          >MEME format</a> or the HTML (<code>.html</code>) or plain text (<code>.txt</code>) 
          output of MEME or DREME. Each of these motifs will be searched against the
          target databases. If you only wish to search with a subset of these
          motifs then look into the <span class="popt">-m</span> and
          <span class="popt">-mi</span> options.</p>
        </div>
        <h4>Target Motif Databases</h4>
        <div class="indent">
          <p>File(s) containing <a href="meme-format.html">MEME formatted motifs</a>.
          Outputs from MEME and DREME are supported, as well as Minimal MEME
          Format. You can convert many other motif formats to MEME format
          using <a href="overview.html#motif_conversion_utilities">conversion scripts</a>
          available with the MEME Suite.</p>
        </div>
      </div>

      <h3>Output</h3>
      <div class="indent">
        <p>Tomtom writes its output to files in a directory named
        <code>tomtom_out</code>, which it creates if necessary. (You can also
        cause the output to be written to a different directory; see
        <span class="popt">-o</span> and <span class="popt">-oc</span>, below.)</p>

        <p>The main output file is 
	named <a href=examples/tomtom_example_output_files/tomtom.html><code>tomtom.html</code></a> 
	and can be viewed with a web browser. The <code>tomtom.html</code> file is created from the
        <a href=examples/tomtom_example_output_files/tomtom.xml><code>tomtom.xml</code></a> file. 
	An additional file, 
        <a href=examples/tomtom_example_output_files/tomtom.txt><code>tomtom.txt</code></a>,
        contains a simplified, text-only version of the output. (See
        <code>-text</code>, below, for the text output format.)</p>
        
        <p>For each query-target match, two additional files containing LOGO
        alignments may also be written -- an encapsulated postscript file
        (<code>.eps</code>) if the <span class="popt">-eps</span> flag is
        specified and a portable network graphic file (<code>.png</code>) if the
        <span class="popt">-png</span> flag is specified. An install of
        ghostscript is required to create the png file.</p>

        <p>Only matches for which the significance is less than or equal to the
        threshold set by the <span class="popt">-thresh</span> switch will be
        shown. By default, significance is measured by <i>q</i>-value of the
        match. The <i>q</i>-value is the estimated false discovery rate if the
        occurrence is accepted as significant. See Storey JD, Tibshirani R,
        "Statistical significance for genome-wide studies". 
        <i>Proc. Natl Acad. Sci. USA (2003) 100:9440&ndash;9445</i></p>
      </div>

        <h3>Options</h3>
      <div class="indent">
        <table class="options">
          <thead>
            <tr>
              <th>Option</th><th>Parameter</th><th>Description</th><th>Default Behaviour</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="4">Input</th>
            </tr>
            <tr>
              <td class="popt">-m</td><td><span class="pdat">id</span></td>
              <td>The name of a motif in the query file that will be used. 
              This option may be repeated multiple times.</td>
              <td>If both this option and the related <span class="popt">-mi</span>
              is unused then all motifs in the query file will be used.</td>
            </tr>
            <tr>
              <td class="popt">-mi</td><td><span class="pdat">index</span></td>
              <td>The offset in the query file of a motif that will be used. This
              option may be repeated multple times.</td>
              <td> If both this option and the related <span class="popt">-m</span>
              is unused then all motifs in the query file will be used.</td>
            </tr>
            <tr>
              <td class="popt">-bfile</td>
              <td><span class="pdat">file</span></td>
              <td>Specify the source of a background model for converting a frequency matrix to 
		a log-odds score matrix.  The background model contains letter frequencies
		that can adjust for biases in the single-letter composition of the sequences
		in the organism from which the motif is derived.
		The <span class="pdat">file</span> must be
		in <a href=bfile-format.html>Markov Background Model Format</a>.
              </td>
              <td>Background frequencies will be loaded from the query motif file.
              Note this was changed from previous versions that used the first
              target motif database because that design choice could not work
              with the <span class="popt">-xalph</span> option.</td>
            </tr>
            <tr>
              <td class="popt">-motif-pseudo</td><td><span class="pdat">count</span></td>
              <td>This option adds the specified pseudocount to the motifs.
                The pseudocount is distributed taking into account the
                background.
                Note that some comparison algorithms require motif probabilities
                to not contain zeros. If you choose to set a pseudocount of
                zero with those comparison algorithms then any motifs
                containing a probability of zero will be discarded and
                a warning will be emitted.</td>
              <td>A pseudocount of 0.1 is added to the motifs.</td>
            </tr>
	    <script>
              wropt("all-xalph", {ddash: false, "tomtom": true, "nospan": true});
	    </script>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Output</th>
            </tr>
            <script>
              wropt("all-o", {"dir": "tomtom_out"});
              wropt("all-oc", {"dir": "tomtom_out"});
            </script>
            <tr>
              <td class="popt">-png</td><td>&nbsp;</td>
              <td>Output motif logo alignment images in portable network graphics
              (png) format. This format is useful for display on websites.</td>
              <td>Images are not output in png format.</td>
            </tr>
            <tr>
              <td class="popt">-eps</td><td>&nbsp;</td>
              <td>Output motif logo alignment images in Encapsulated Postscript
              (eps) format. This format is useful for inclusion in publications as
              it is a vector graphics format and can be easily scaled.</td>
              <td>Images are not output in eps format.</td>
            </tr>
            <tr>
              <td class="popt">-text</td><td>&nbsp;</td>
              <td>This option causes Tomtom to print just a tab-delimited text
              file to standard output. The output begins with a header, indicated
              by leading "#" characters. This is followed by a single title line,
              and then the actual values. The columns are
              <table class="light">
                <thead>
                  <tr><th>Column</th><th>Contents</th></tr>
                </thead>
                <tbody>
                  <tr><td>1</td><td>Query motif name</td></tr>
                  <tr><td>2</td><td>Target motif name</td></tr>
                  <tr><td>3</td><td>Optimal offset: the offset between the query and the target motif</td></tr>
                  <tr><td>4</td><td><i>p</i>-value</td></tr>
                  <tr><td>5</td><td><i>E</i>-value</td></tr>
                  <tr><td>6</td><td><i>q</i>-value</td></tr>
                  <tr><td>7</td><td>Overlap: the number of positions of overlap between the two motifs.</td></tr>
                  <tr><td>8</td><td>Query consensus sequence.</td></tr>
                  <tr><td>9</td><td>Target consensus sequence.</td></tr>
                  <tr><td>10</td><td>Orientation: Orientation of target motif with respect to query motif.</td></tr>
                </tbody>
              </table>
              </td>
              <td>The program runs as normal.</td>
            </tr>
            <tr>
              <td class="popt">-no-ssc</td><td>&nbsp;</td>
              <td>This option causes the LOGOs in the LOGO alignments output by
              Tomtom not to be corrected for small-sample sizes. By default, the
              height of letters in the LOGOs are reduced when the number of
              samples on which a motif is based (<code>nsites</code> in the MEME
              motif) is small. The default setting can cause motifs based on very
              few sites to have "empty" LOGOs, so this switch can be used if your
              query or target motifs are based on few samples.</td>
              <td>Small sample correction is used.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Scoring</th>
            </tr>
            <tr>
              <td class="popt">-incomplete-scores</td><td>&nbsp;</td>
              <td>Compute scores using only aligned columns.</td>
              <td>Take into account columns that don't align.</td>
            </tr>
            <tr>
              <td class="popt">-thresh</td><td><span class="pdat">value</span></td>
              <td>Only report matches with significance values &le; 
              <span class="pdat">value</span>. Unless the <code>-evalue</code>
              option is specifed then this value must be smaller than or equal to
              1.</td>
              <td>A threshold of 0.5 is used.</td>
            </tr>
            <tr>
              <td class="popt">-evalue</td><td>&nbsp;</td>
              <td>Use the <i>E</i>-value of the match as the significance
              threshold</td>
              <td>Use the <i>q</i>-value as the significance threshold</td>
            </tr>
            <tr>
              <td class="popt">-dist</td>
              <td>allr|&#8203;ed|&#8203;kullback|&#8203;pearson|&#8203;sandelin</td>
              <td>
                <table class="dark">
                  <thead>
                    <tr>
                      <th>Code</th><th>Name</th><th>Restrictions</th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>allr</td>
                      <td>Average log-likelihood ratio</td>
                      <td>Non-zero probabilities</td>
                    </tr>
                    <tr>
                      <td>ed</td>
                      <td>Euclidian distance</td>
                      <td></td>
                    </tr>
                    <tr>
                      <td>kullback</td>
                      <td>Kullback-Leibler divergence</td>
                      <td></td>
                    </tr>
                    <tr>
                      <td>pearson</td>
                      <td>Pearson correlation coefficient</td>
                      <td></td>
                    </tr>
                    <tr>
                      <td>sandelin</td>
                      <td>Sandelin-Wasserman function</td>
                      <td></td>
                    </tr>
                    <tr>
                      <td>blic1</td>
                      <td>Bayesian Likelihood 2-Components (1 Dirichlet)</td>
                      <td>DNA only</td>
                    </tr>
                    <tr>
                      <td>blic5</td>
                      <td>Bayesian Likelihood 2-Components (5 Dirichlet)</td>
                      <td>DNA only</td>
                    </tr>
                    <tr>
                      <td>llr1</td>
                      <td>Log likelihood ratio (1 Dirichlet)</td>
                      <td>DNA only</td>
                    </tr>
                    <tr>
                      <td>llr5</td>
                      <td>Log likelihood ratio (5 Dirichlet)</td>
                      <td>DNA only</td>
                    </tr>
                  </tbody>
                </table>
                Detailed descriptions of these functions can be found in the published description of
                <a href="http://noble.gs.washington.edu/papers/gupta_quantifying.pdf">Tomtom</a>.
              </td>
              <td>Pearson correlation coefficient is used by default.</td>
            </tr>
            <tr>
              <td class="popt">-internal</td><td>&nbsp;</td>
              <td>This parameter forces the shorter motif to be completely
              contained in the longer motif.</td>
              <td>The shorter motif may extend outside the longer motif.</td>
            </tr>
            <tr>
              <td class="popt">-min-overlap</td><td><span class="pdat">min overlap</span></td>
              <td>Only report motif matches that overlap by
              <span class="pdat">min overlap</span> positions or more. In case a
              query motif is smaller than <span class="pdat">min overlap</span>,
              then the motif's width is used as the minimum overlap for that
              query.</td>
              <td>A minimum overlap of 1 is required.</td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th colspan="4">Miscellaneous</th>
            </tr>
            <script>wropt("all-verbosity");</script>
            <tr>
              <td class="popt">-version</td><td></td>
              <td>Display the version and exit.</td>
              <td>Run as normal.</td>
            </tr>
          </tbody>
        </table>
      </div>

    </div>

    <h3>Citing</h3>
    <div class="indent">
      <p>
      If you use Tomtom in your research, please cite the following paper:<br />
      <span class="citation">
        Shobhit Gupta, JA Stamatoyannopolous, Timothy Bailey and William Stafford Noble,
        &quot;Quantifying similarity between motifs&quot;,
        <i>Genome Biology</i>, <b>8</b>(2):R24, 2007.
        <a href="http://genomebiology.com/2007/8/2/R24">[full text]</a>
      </span>
      </p>
    </div>

    <script src="js/manual.js"></script>

  </body>
</html>

